internal/users/localentries/localgroups.go

```diff
@@ -116,140 +226,142 @@ func existingLocalGroups(user, groupPath string) (groups []string, err error) {
-	existingUsers := make(map[string]struct{})
-	usernames, err := opts.getUsersFunc()
-	if err != nil {
+	if err := types.ValidateGroupEntries(groups); err != nil {
 		return err
 	}
-	for _, username := range usernames {
-		existingUsers[username] = struct{}{}
-	}
-	// If no username was returned, something probably went wrong during the getpwent call and we should stop,
-	// otherwise we would remove all users from the local groups.
-	if len(existingUsers) == 0 {
-		return errors.New("no existing users found, local groups won't be cleaned")
-	}
 
-	// Get the list of local groups
-	f, err := os.Open(opts.groupPath)
-	if err != nil {
-		return err
+	backupPath := groupFileBackupPath(groupPath)
+	oldBackup := ""
+	groupsEntries := formatGroupEntries(groups)
+
+	log.Debugf(context.TODO(), "Saving group entries %#v to %q", groups, groupPath)
+	if len(groupsEntries) > 0 {
+		log.Debugf(context.TODO(), "Group file content:\n%s", groupsEntries)
 	}
-	defer f.Close()
 
-	// Format of a line composing the group file is:
-	// group_name:password:group_id:user1,…,usern
-	var delOps [][]string
-	scanner := bufio.NewScanner(f)
-	for scanner.Scan() {
-		t := strings.TrimSpace(scanner.Text())
-		if t == "" {
-			continue
+	if tmpDir, err := os.MkdirTemp(os.TempDir(), "authd-groups-backup"); err == nil {
+		defer os.Remove(tmpDir)
+
+		b := filepath.Join(tmpDir, filepath.Base(backupPath))
+		err := fileutils.CopyFile(backupPath, b)
+		if err == nil {
+			log.Debugf(context.Background(), "Backup of %q saved to %q", backupPath, b)
+			oldBackup = b
+			defer os.Remove(oldBackup)
 		}
-		elems := strings.Split(t, ":")
-		if len(elems) != 4 {
-			return fmt.Errorf("malformed entry in group file (should have 4 separators): %q", t)
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
+			log.Warningf(context.Background(), "Failed to create backup of %q: %v",
+				backupPath, err)
 		}
+	}
 
-		groupName := elems[0]
-		users := strings.Split(elems[3], ",")
-		for _, user := range users {
-			if _, ok := existingUsers[user]; ok {
-				continue
-			}
+	if err := os.Remove(backupPath); err != nil && !errors.Is(err, os.ErrNotExist) {
+		log.Warningf(context.Background(), "Failed to remove group file backup: %v", err)
+	}
 
-			// User doesn't exist anymore, remove it from the group
-			args := opts.gpasswdCmd[1:]
-			delOps = append(delOps, append(args, "--delete", user, groupName))
+	log.Debugf(context.Background(), "Backing up %q to %q", inputPath, backupPath)
+	backupAction := os.Rename
+	if fi, _ := os.Lstat(inputPath); fi != nil && fi.Mode()&fs.ModeSymlink != 0 {
+		backupAction = fileutils.CopyFile
+	}
+	if err := backupAction(inputPath, backupPath); err != nil {
+		log.Warningf(context.Background(), "Failed make a backup for the group file: %v", err)
+
+		if oldBackup != "" {
+			// Backup of current group file failed, let's restore the old backup.
+			if err := fileutils.Lrename(oldBackup, backupPath); err != nil {
```

Comment:
the old backup can be on a different filesystem, which can cause the rename syscall to fail

Reply:
Yeah... I actually have the issue in my container, where authd is mounted.... However I just kept the code we had before here, it's just the copy from the migration.

I didn't bother fixing it here too, but maybe it's worth adding an issue?